### PR TITLE
ENH: Functions to calculate the average number of reflections and their density

### PR DESCRIPTION
**Pull request by mberz**
### Changes proposed in this pull request:

- Add functions to calculate the average
  - reflection rate in a room based on diffuse field assumptions
  - total number of reflections in a diffuse field

The PR is partially inspired by https://github.com/pyfar/pyfar/pull/782 by @ahms5, but implements the reflection density and total number of reflections separately from RIR synthesis.

### CODE REVIEW

File: tests/test_parametric.py [hunk 1]
```diff
@@ -6,6 +6,7 @@
 import pyrato.parametric as parametric
 from pyrato.parametric import mean_free_path
 import pyrato as ra
+import pyrato
```
[the remark this answers is not in the document]
Comment (mberz) [reply tests/test_parametric.py:9]: Importing as ra does not follow our conventions and should be replaced in the future.

File: pyrato/parametric.py [hunk 2]
```diff
@@ -302,3 +302,142 @@ def reverberation_time_sabine(
...
+
+
+def average_reflection_density(
+        volume: float,
+        times: np.ndarray | List,
+        speed_of_sound: float | None = None,
+    ) -> pf.TimeData:
+    r"""Calculate the time dependent average reflection density in a room.
+
+    The reflection density is calculated as the following ratio [#]_
```
Comment (hoyer-a) [inline pyrato/parametric.py]: Could you add the equation number (eq. 4.5) or the subsection (sec. 4.2)? It can get annoying to search for equations in books.

File: pyrato/parametric.py [hunk 3]
```diff
@@ -302,3 +302,142 @@ def reverberation_time_sabine(
...
+
+    Parameters
+    ----------
+    volume : float
+        Volume of the room :math:`V` in :math:`m^3`.
+    times : numpy.ndarray, list
+        Time vector in seconds.
+    speed_of_sound : float, None, optional
+        Speed of sound in the room. By default, the
+        :py:attr:`~pyfar.constants.reference_speed_of_sound` is used.
```
Comment (hoyer-a) [inline pyrato/parametric.py]: this should do the trick

```suggestion
        :py:data:`~pyfar.constants.reference_speed_of_sound` is used.
```

File: pyrato/parametric.py [hunk 4]
```diff
@@ -302,3 +302,142 @@ def reverberation_time_sabine(
...
```
Comment (hoyer-a) [inline pyrato/parametric.py]:
```suggestion
        :py:data:`~pyfar.constants.reference_speed_of_sound` is used.
```

File: pyrato/parametric.py [hunk 5]
```diff
@@ -302,3 +302,142 @@ def reverberation_time_sabine(
...
+    speed of sound in the room, and :math:`t` is the time vector in seconds.
+
+    Parameters
+    ----------
+    volume : float
+        Volume of the room :math:`V` in :math:`m^3`.
+    times : numpy.ndarray, list
+        Time vector in seconds.
+    speed_of_sound : float, None, optional
+        Speed of sound in the room. By default, the
```
Comment (hoyer-a) [inline pyrato/parametric.py]:
```suggestion
        Speed of sound in the room. By default (`None`), the
```

File: pyrato/parametric.py [hunk 6]
```diff
@@ -302,3 +302,142 @@ def reverberation_time_sabine(
...
```
Comment (hoyer-a) [inline pyrato/parametric.py]:
```suggestion
        Speed of sound in the room. By default (`None`), the
```

File: pyrato/parametric.py [hunk 7]
```diff
@@ -302,3 +302,142 @@ def reverberation_time_sabine(
...
+
+
+def average_number_of_reflections(
+        volume: float,
+        times: np.ndarray | List,
+        speed_of_sound: float | None = None,
+    ) -> pf.TimeData:
+    r"""Calculate the time dependent average number of reflections in a room.
+
+    The average number of reflections is calculated as the following ratio [#]_
```
Comment (hoyer-a) [inline pyrato/parametric.py]: same here, maybe reference sec 4.2

**Review by hoyer-a — COMMENTED** [review]
Thanks, just some small comments on documentation

File: pyrato/parametric.py [hunk 8]
```diff
@@ -302,3 +302,142 @@ def reverberation_time_sabine(
...
 
     return reverberation_time
+
+
+def average_reflection_density(
+        volume: float,
+        times: np.ndarray | List,
+        speed_of_sound: float | None = None,
+    ) -> pf.TimeData:
+    r"""Calculate the time dependent average reflection density in a room.
```
Comment (artur-pa) [inline pyrato/parametric.py:312]: Maybe we should mention that this expression is derived for a diffusely reflecting rectangular room and therefore relies on the diffuse-field assumption. Or do you think the reference is sufficient here?

File: pyrato/parametric.py [hunk 9]
```diff
@@ -302,3 +302,142 @@ def reverberation_time_sabine(
...
+    ) -> pf.TimeData:
+    r"""Calculate the time dependent average number of reflections in a room.
+
+    The average number of reflections is calculated as the following ratio [#]_
+
+    .. math::
+        N(t) = \frac{4 \pi c^3 t^3}{3 V},
+
+    where :math:`V` is the room volume in :math:`m^3`, :math:`c` is the
+    speed of sound in the room, and :math:`t` is the time vector in seconds.
```
Comment (artur-pa) [inline pyrato/parametric.py:396]: Maybe here also ...

>  The equation is derived under the diffuse-field assumption. [reference]. ?

File: pyrato/parametric.py [hunk 10]
```diff
@@ -302,3 +302,142 @@ def reverberation_time_sabine(
...
+        volume: float,
+        times: np.ndarray | List,
+        speed_of_sound: float | None = None,
+    ) -> pf.TimeData:
+    r"""Calculate the time dependent average number of reflections in a room.
+
+    The average number of reflections is calculated as the following ratio [#]_
+
+    .. math::
+        N(t) = \frac{4 \pi c^3 t^3}{3 V},
```
Comment (artur-pa) [inline pyrato/parametric.py:393]: Maybe we should mention that the average number of reflections is derived
from the average reflection density in our method:

```
.. math::
    N(t) = \frac{4 \pi c^3 t^3}{3V} =
    \frac{\mu(t)t}{3},
```

where` :math:`\mu(t)`` is the average reflection density obtained from
`:func:`average_reflection_density``.

File: tests/test_parametric.py [hunk 11]
```diff
@@ -98,3 +99,101 @@ def test_sabine_zero_absorption_area():
...
+    np.testing.assert_allclose(
+        np.squeeze(density.time),
+        reference,
+    )
+
+
+def test_reflection_density_errors():
+    """Test if all errors are raised correctly."""
+
+    with pytest.raises(ValueError, match="must be positive"):
```
Comment (artur-pa) [inline tests/test_parametric.py:133]: Should we also add a test for volume = 0? Since the implementation checks volume <= 0, it would be good to cover the boundary case in addition to a negative value.

File: pyrato/parametric.py (hunk 10 above)
Comment (mberz) [reply pyrato/parametric.py:393]: I'd say it's more the other way around (the density is the derivative of the total number). Anyhow, I'd not add this here.

File: pyrato/parametric.py (hunk 8 above)
Comment (mberz) [reply pyrato/parametric.py:312]: The expressions are derived using the image source method. There's no need for diffuse field assumptions. I'll add the chapter to the reference as @hoyer-a suggested.

**Review by hoyer-a — APPROVED** [review]
